### PR TITLE
sources: add source handler integration

### PR DESCRIPTION
**Pull request by cmatsuoka**
Provide calls to integrate source handlers with state manager and the
executor, and use it to check if sources are outdated during the
planning phase.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_parts/sources/errors.py [hunk 1]
```diff
@@ -16,13 +16,57 @@
...
+            f"Failed to pull source: {option!r} cannot be used "
+            f"with a {source_type} source."
+        )
+        resolution = "Make sure sources are correctly specified."
+
+        super().__init__(brief=brief, resolution=resolution)
+
+
+class IncompatibleSourceOptions(SourceError):
+    """Source specified options that can't be used at the same time."""
```
Comment (sergiusens) [inline craft_parts/sources/errors.py:55]:
```suggestion
    """Source specified options that cannot be used at the same time."""
```

File: craft_parts/sources/sources.py [hunk 2]
```diff
@@ -71,16 +71,113 @@
...
+        dirs = ProjectDirs()
+
+    source_handler = None
+    if part.spec.source:
+        handler_class = _get_source_handler_class(
+            part.spec.source,
+            source_type=part.spec.source_type,
+        )
+        source_handler = handler_class(
+            application_name=application_name,
```
Comment (sergiusens) [inline craft_parts/sources/sources.py:118]: refresh my memory on why the application name is important here

File: craft_parts/sources/sources.py [hunk 3]
```diff
@@ -71,16 +71,113 @@
...
+            source=part.spec.source,
+            part_src_dir=part.part_src_dir,
+            source_checksum=part.spec.source_checksum,
+            source_branch=part.spec.source_branch,
+            source_tag=part.spec.source_tag,
+            source_depth=part.spec.source_depth,
+            source_commit=part.spec.source_commit,
+            dirs=dirs,
```
Comment (sergiusens) [inline craft_parts/sources/sources.py:126]: looking at this, it seems that we need to make this var a bit more clear on why it's there with a better name

File: craft_parts/sources/sources.py [hunk 4]
```diff
@@ -71,16 +71,113 @@
...
+        )
+
+    return source_handler
+
+
+def _get_source_handler_class(source, *, source_type: str = "") -> SourceHandlerType:
```
Comment (sergiusens) [inline craft_parts/sources/sources.py:132]: on load is source_type None or ""?

Comment (cmatsuoka) [reply craft_parts/sources/sources.py:132]: It's an empty string in the unmarshaled part if not specified.

File: craft_parts/sources/sources.py (hunk 2 above)
Comment (cmatsuoka) [reply craft_parts/sources/sources.py:118]: It's used to compose the path to the downloaded file cache.

File: craft_parts/sources/sources.py (hunk 3 above)
Comment (cmatsuoka) [reply craft_parts/sources/sources.py:126]: Renamed it to `project_dirs` to emphasize the relation to the `ProjectDirs` class.

**Review by cjp256 — APPROVED** [review]

File: craft_parts/sources/sources.py (hunk 2 above)
Comment (sergiusens) [reply craft_parts/sources/sources.py:115]: Ah, maybe this can be global, or in other words, scoped to craft-parts and not specific applications.

**Review by sergiusens — APPROVED** [review]
Looks good, we might want to reassess the use of application for sources.

**Pull request merged**